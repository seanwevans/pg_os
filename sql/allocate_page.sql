-- regression tests for allocate_page
\set ECHO none
SET client_min_messages TO warning;
DROP EXTENSION IF EXISTS pg_os CASCADE;
CREATE EXTENSION pg_os;

-- prepare memory pages and a thread that will allocate them
INSERT INTO pages (size)
SELECT 4096 FROM generate_series(1, 20);

SELECT create_user('allocator') AS user_id \gset
INSERT INTO processes (name, state, owner_user_id)
VALUES ('alloc_proc', 'new', :user_id)
RETURNING id AS process_id \gset
INSERT INTO threads (process_id, name)
VALUES (:process_id, 'alloc_thread')
RETURNING id AS thread_id \gset

\set ECHO queries
\set VERBOSITY terse

-- allocate a batch of pages for the same thread
SELECT allocate_page(:thread_id) FROM generate_series(1, 10);

-- verify the virtual addresses were allocated monotonically
SELECT virtual_address
  FROM page_tables
 WHERE thread_id = :thread_id
 ORDER BY virtual_address;
