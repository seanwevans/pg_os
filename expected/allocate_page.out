-- regression tests for allocate_page
\set ECHO none
SELECT allocate_page(1) FROM generate_series(1, 10);
 allocate_page 
---------------
          4096
          8192
         12288
         16384
         20480
         24576
         28672
         32768
         36864
         40960
(10 rows)

SELECT virtual_address
  FROM page_tables
 WHERE thread_id = 1
 ORDER BY virtual_address;
 virtual_address 
-----------------
            4096
            8192
           12288
           16384
           20480
           24576
           28672
           32768
           36864
           40960
(10 rows)

